File: admin/registration_code.go
package admin

import (
	"database/sql"
	"net/http"
	"strconv"

	"neoai/auth"
	"neoai/utils"

	"github.com/gin-gonic/gin"
)

// registration_code.go — admin endpoints for managing registration codes.
//
//   GET    /admin/registration-code/list      — list all codes
//   POST   /admin/registration-code/generate   — generate N new codes
//   GET    /admin/registration-code/state     — is registration code required?
//   POST   /admin/registration-code/state     — toggle require_code flag
//   POST   /admin/registration-code/disable   — disable a code
//   GET    /admin/registration-code/delete/:id — delete a code

type GenerateRegCodeForm struct {
	Number  int     `json:"number"`
	Quota   float32 `json:"quota"`
	MaxUses int     `json:"max_uses"`
	Note    string  `json:"note"`
}

type RegCodeStateForm struct {
	Required bool `json:"required"`
}

type RegCodeDisableForm struct {
	Id int64 `json:"id"`
}

func ListRegCodesAPI(c *gin.Context) {
	db := utils.GetDBFromContext(c)
	codes, err := auth.ListRegistrationCodes(db)
	if err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status": false,
			"error":  err.Error(),
		})
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"status": true,
		"data":   codes,
	})
}

func GenerateRegCodesAPI(c *gin.Context) {
	db := utils.GetDBFromContext(c)
	var form GenerateRegCodeForm
	if err := c.ShouldBindJSON(&form); err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status":  false,
			"message": err.Error(),
		})
		return
	}
	if form.Number <= 0 || form.Number > 1000 {
		c.JSON(http.StatusOK, gin.H{
			"status":  false,
			"message": "number must be between 1 and 1000",
		})
		return
	}
	if form.MaxUses <= 0 {
		form.MaxUses = 1
	}
	codes, err := auth.GenerateRegistrationCodes(db, form.Number, form.Quota, form.MaxUses, form.Note)
	if err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status":  false,
			"message": err.Error(),
		})
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"status": true,
		"data":   codes,
	})
}

func GetRegCodeStateAPI(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{
		"status":   true,
		"required": auth.IsRegistrationCodeRequired(),
	})
}

func SetRegCodeStateAPI(c *gin.Context) {
	var form RegCodeStateForm
	if err := c.ShouldBindJSON(&form); err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status":  false,
			"message": err.Error(),
		})
		return
	}
	// Persist to config.yaml under system.registration.require_code
	if err := utils.SaveConfig("system.registration.require_code", form.Required); err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status":  false,
			"message": err.Error(),
		})
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"status":   true,
		"required": form.Required,
	})
}

func DisableRegCodeAPI(c *gin.Context) {
	db := utils.GetDBFromContext(c)
	var form RegCodeDisableForm
	if err := c.ShouldBindJSON(&form); err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status":  false,
			"message": err.Error(),
		})
		return
	}
	if err := auth.DisableRegistrationCode(db, form.Id); err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status":  false,
			"message": err.Error(),
		})
		return
	}
	c.JSON(http.StatusOK, gin.H{"status": true})
}

func DeleteRegCodeAPI(c *gin.Context) {
	db := utils.GetDBFromContext(c)
	id, _ := strconv.ParseInt(c.Param("id"), 10, 64)
	if err := auth.DeleteRegistrationCode(db, id); err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status":  false,
			"message": err.Error(),
		})
		return
	}
	c.JSON(http.StatusOK, gin.H{"status": true})
}

// silence unused import warning (sql is referenced by auth indirectly)
var _ = sql.ErrNoRows

File: app/src/admin/api/registration-code.ts
import axios from "axios";
import { getErrorMessage } from "@/utils/base.ts";
import { CommonResponse } from "@/api/common.ts";

// Registration code admin API client.

export type RegistrationCode = {
  id: number;
  code: string;
  quota: number;
  max_uses: number;
  used_count: number;
  note: string;
  created_at: string;
  disabled: boolean;
};

export type RegistrationCodeListResponse = CommonResponse & {
  data: RegistrationCode[];
};

export type GenerateRegCodeForm = {
  number: number;
  quota: number;
  max_uses: number;
  note: string;
};

export type GenerateRegCodeResponse = CommonResponse & {
  data: string[];
};

export type RegCodeStateResponse = CommonResponse & {
  required: boolean;
};

export async function listRegistrationCodes(): Promise<RegistrationCodeListResponse> {
  try {
    const r = await axios.get("/admin/registration-code/list");
    return r.data as RegistrationCodeListResponse;
  } catch (e) {
    return { status: false, error: getErrorMessage(e), data: [] };
  }
}

export async function generateRegistrationCodes(
  form: GenerateRegCodeForm,
): Promise<GenerateRegCodeResponse> {
  try {
    const r = await axios.post("/admin/registration-code/generate", form);
    return r.data as GenerateRegCodeResponse;
  } catch (e) {
    return { status: false, error: getErrorMessage(e), data: [] };
  }
}

export async function getRegistrationCodeState(): Promise<RegCodeStateResponse> {
  try {
    const r = await axios.get("/admin/registration-code/state");
    return r.data as RegCodeStateResponse;
  } catch (e) {
    return { status: false, error: getErrorMessage(e), required: false };
  }
}

export async function setRegistrationCodeState(
  required: boolean,
): Promise<RegCodeStateResponse> {
  try {
    const r = await axios.post("/admin/registration-code/state", { required });
    return r.data as RegCodeStateResponse;
  } catch (e) {
    return { status: false, error: getErrorMessage(e), required };
  }
}

export async function disableRegistrationCode(id: number): Promise<CommonResponse> {
  try {
    const r = await axios.post("/admin/registration-code/disable", { id });
    return r.data as CommonResponse;
  } catch (e) {
    return { status: false, error: getErrorMessage(e) };
  }
}

export async function deleteRegistrationCode(
  id: number,
): Promise<CommonResponse> {
  try {
    const r = await axios.get(`/admin/registration-code/delete/${id}`);
    return r.data as CommonResponse;
  } catch (e) {
    return { status: false, error: getErrorMessage(e) };
  }
}
